### DISCUSSION TITLE
Package names that end with a semver2-compliant version string

### DISCUSSION THREAD

**Discussion by mmraff** [npm/feedback#507] [Registry]
I'd like to know if your team has worked out a solution to this, because it seems to me that even the new rules for package naming (so helpfully implemented in package validate-npm-package-name) are loose enough to allow some major ambiguity, in particular when it comes to naming files in the npm registry.

1. A package name can include dashes, dots, and digits, with no restriction on the arrangement (except no name-initial dot);
2. So a package name can end with something that has the form [number].[number].[number];
3. In fact there can be a hyphen-chained series of such strings on the end of the name;
4. The Semantic Versioning 2.0 spec also allows a pre-release string to be of the form [number].[number].[number];
5. Pre-release versions of packages can be published to the registry;
6. There can be a package in the registry with the name "name", and simultaneously there can be a package in the registry with the name "name-1.2.3"

So how do you resolve the registry URL for "name@1.2.3-4.5.6" versus "name-1.2.3@4.5.6"?
And that's just one of the simplest examples of what's bugging me...

**Comment by ljharb** [comment]
You can't have an "@" in a package name except at the beginning of a scope - so the first one would be "name" at v1.2.3-4.5.6, and the second "name-1.2.3" at v4.5.6.

**Reply by mmraff** [reply]
Sorry, it seems I wasn't clear enough.
I was talking about the filenames in registry URLs, i.e. [name]/-/[name]-[version].tgz (or @[scope]/[name]/-/[name]-[version].tgz)
I know that this is not the kind of thing that affects the npm CLI, but it seems to me that there must be the generation of a filename somewhere in the chain between publishing and end-user downloading.
It might come off as a totally arcane concern, but I'm developing a module that works with the names of package tarballs.
Ultimately I may decide to impose my own rules on such naming, since I have control over saving downloaded tarballs or creating them on my side; but I'd prefer to conform to npm practice - if there's actually an issue here that npm architecture would need to address.

**Reply by ljharb** [reply]
I guess I'm still not sure where the issue lies. In those URLs, a name can only start with an @, if it's scoped, otherwise it can't have one, and since the `-` and the version must always be present, it seems unambiguously parseable.

**Reply by mmraff** [reply]
The '@' is irrelevant to the problem, because this is not about a spec passed to npm-package-arg.
Consider a registry URL that ends with
/name-1.2.3-4.5.6.tgz
Does that correspond to name v1.2.3 pre-release 4.5.6, or name-1.2.3 v4.5.6?

**Reply by mmraff** [reply]
BTW, when I say "registry URL", I'm talking about the _resolved field in the manifest, such as you would obtain from calling pacote.manifest(spec)

**Reply by ljharb** [reply]
given that the last dash separates "4.5.6" from "name-1.2.3", that to me is obviously v4.5.6 of a package named "name-1.2.3".

**Reply by mmraff** [reply]
Would that be because the pre-release string is never made part of the filename?

**Reply by ljharb** [reply]
ohhh, i see what you're asking now, thanks for bearing with me :-)

You're saying that in addition to "name-1.2.3" `v4.5.6`, there could also be "name" `v1.2.3-4.5.6`.

Does npm actually allow the full semver2 prerelease range? npm predates semver 2, iirc.

**Reply by mmraff** [reply]
Now THERE'S the question then. It would bring me relief to find out that the npm registry restricts the publishable range of pre-release strings even further than the semver2 specification, because then there's light at the end of the tunnel for my quest.
But I'll believe it when I see it documented...

**Reply by ethomson** [reply]
> Consider a registry URL that ends with
/name-1.2.3-4.5.6.tgz
Does that correspond to name v1.2.3 pre-release 4.5.6, or name-1.2.3 v4.5.6?

You do not have enough information from the filename alone to answer that question.  You would need to look at the packument, or even the full URL (which would have the package name in it).

Consider the `@ethomson/version-test-1.2.3` package at version `4.5.6-7.8.9`: [http://registry.npmjs.com/@ethomson/version-test-1.2.3](http://registry.npmjs.com/@ethomson/version-test-1.2.3).  You can see that its full tarball URL is `https://registry.npmjs.org/@ethomson/version-test-1.2.3/-/version-test-1.2.3-4.5.6-7.8.9.tgz`.

You'd need the full registry URL to understand whether this is package `version-test-1.2.3` at version `4.5.6-7.8.9` or `version-test-1.2.3-4.5.6` at version `7.8.9` [http://registry.npmjs.com/@ethomson/version-test-1.2.3-4.5.6](http://registry.npmjs.com/@ethomson/version-test-1.2.3-4.5.6).  (Or to read the packument - but you cannot tell by filename alone.)

**Reply by ljharb** [reply]
@ethomson what happens if both package/version combos exist though? wouldn't the same URL end up referencing two package/version combos?

**Reply by ethomson** [reply]
No (if I'm understanding the question correctly) because the URL has the package in it.  (You can see in the above example.)

`@ethomson/version-test-1.2.3` version `4.5.6-7.8.9` is URL: `https://registry.npmjs.org/@ethomson/version-test-1.2.3/-/version-test-1.2.3-4.5.6-7.8.9.tgz`
`@ethomson/version-test-1.2.3-4.5.6` version `7.8.9` is URL: `https://registry.npmjs.org/@ethomson/version-test-1.2.3-4.5.6/-/version-test-1.2.3-4.5.6-7.8.9.tgz`

**Reply by ljharb** [reply]
In that case, you can indeed parse the package name out of the URL, and thus in the tarball name, figure out which part is the version and which is the package name - it's only if you have *just* the tarball that you have to extract it to determine all that?

**Comment by mmraff** [comment]
> You do not have enough information from the filename alone to answer that
question. You would need to look at the packument, or even the full URL...

Thanks, that, with the registry examples you gave, is a perfectly
satisfying answer. :-)
(Also a letdown. Oh well.)
